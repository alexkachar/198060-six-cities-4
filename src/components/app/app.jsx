import React from "react";
import PropTypes from "prop-types";
import {BrowserRouter, Route, Switch} from "react-router-dom";

import Main from "../main/main.jsx";
import OfferDetails from "../offer-details/offer-details.jsx";
import {OfferTypes} from '../../types/types.js';

class App extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      selectedOffer: null
    };

    this.handleTitleClick = this.handleTitleClick.bind(this);
  }

  handleTitleClick(offer) {
    this.setState({
      selectedOffer: offer,
    });
  }

  _renderApp() {
    const {offers, offersCount} = this.props;
    const {selectedOffer} = this.state;

    return (
      !selectedOffer
        ? <Main offers={offers} offersCount={offersCount} onTitleClick={this.handleTitleClick} />
        : <OfferDetails offer={selectedOffer} />
    );
  }

  render() {
    const {offers} = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._renderApp()}
          </Route>
          <Route exact path="/dev-details">
            <OfferDetails offer={offers[0]} />
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

App.propTypes = {
  offersCount: PropTypes.number.isRequired,
  offers: PropTypes.arrayOf(OfferTypes.isRequired).isRequired
};

export default App;
