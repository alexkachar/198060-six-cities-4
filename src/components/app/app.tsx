import * as React from 'react';
import PropTypes from "prop-types";
import {connect} from 'react-redux';
import {BrowserRouter, Route, Switch} from 'react-router-dom';

import Main from '../main/main';
import Login from '../login/login';
import Favorites from '../favorites/favorites';
import OfferDetails from '../offer-details/offer-details';

import UserOperation from '../../store/operations/user/user';
import DataOperation from '../../store/operations/data/data';
import OfferTypes from '../../types/offer';
import {filterOffers} from '../../store/reducers/filter/selectors';
import {ROUTES} from '../../helpers/constants';
import {getAuthStatus} from '../../store/reducers/user/selectors';
import {AUTH_STATUS} from '../../helpers/constants';
import withPrivateRoute from '../../hocs/with-private-route/with-private-route';
import {getLoadingFlag} from '../../store/reducers/data/selectors';

const App = ({
  offers,
  authStatus,
  loading,
  onLogin,
  onSetFavoriteStatus,
}) => {
  const isAuthorized = authStatus === AUTH_STATUS.auth;
  const FavoritesWrapped = withPrivateRoute(Favorites, isAuthorized);
  return (
    <BrowserRouter>
      <Switch>
        <Route exact path={ROUTES.main}
          render={(props) => {
            return (
              <Main
                {...props}
                offers={offers}
                onSetFavoriteStatus={onSetFavoriteStatus}
              />
            );
          }}
        />
        <Route exact path={ROUTES.details}
          render={({match}) => {
            const {id} = match.params;
            return (
              <OfferDetails
                offerId={id}
                onSetFavoriteStatus={onSetFavoriteStatus}
              />
            );
          }}
        />
        <Route exact path={ROUTES.favorites}
          render={() => {
            return (
              <FavoritesWrapped
                onSetFavoriteStatus={onSetFavoriteStatus}
              />
            );
          }}
        />
        <Route exact path={ROUTES.login}
          render={() => {
            return (
              <Login
                onLogin={onLogin}
                loading={loading}
              />
            );
          }}
        />
      </Switch>
    </BrowserRouter>
  );
};

const mapStateToProps = (state) => ({
  offers: filterOffers(state),
  loading: getLoadingFlag(state),
  authStatus: getAuthStatus(state)
});

const mapDispatchToProps = (dispatch) => ({

  onLogin(authData) {
    dispatch(UserOperation.login(authData));
  },

  onSetFavoriteStatus(offerId, isFavorite) {
    dispatch(DataOperation.setFavoriteStatus(offerId, isFavorite));
  },

});

export {App};

export default connect(mapStateToProps, mapDispatchToProps)(App);

App.propTypes = {
  offers: PropTypes.arrayOf(OfferTypes.isRequired).isRequired,
  authStatus: PropTypes.string.isRequired,
  onLogin: PropTypes.func,
  onSetFavoriteStatus: PropTypes.func.isRequired,
  handleTitleClick: PropTypes.func,
  loading: PropTypes.bool.isRequired,
};
