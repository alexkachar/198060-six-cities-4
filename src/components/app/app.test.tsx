
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import {App} from './app';
import Offer from '../../interfaces/offer';
import mockStore from '../../test-data/mock-store';
import mockOffers from '../../test-data/offers';

const AUTH_STATUS = `NO_AUTH`;
const LOADING = false;
const OFFERS: Offer[] = mockOffers;

it(`App renders correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={mockStore}>
          <App
            offers={OFFERS}
            loading={LOADING}
            authStatus={AUTH_STATUS}
            onLogin={jest.fn()}
            onSetFavoriteStatus={jest.fn()}
            onCardHover={jest.fn()}
            onCardHoverLeave={jest.fn()}
          />
        </Provider>, {
          createNodeMock: () => document.createElement(`div`)
        })
    .toJSON();
  expect(tree).toMatchSnapshot();
});
