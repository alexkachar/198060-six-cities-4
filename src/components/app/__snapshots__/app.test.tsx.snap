// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div
  className="page page--gray page--main"
>
  <header
    className="header"
  >
    <div
      className="container"
    >
      <div
        className="header__wrapper"
      >
        <div
          className="header__left"
        >
          <a
            className="header__logo-link header__logo-link--active"
            href="/"
            onClick={[Function]}
          >
            <img
              alt="6 cities logo"
              className="header__logo"
              height={41}
              src="/img/logo.svg"
              width={81}
            />
          </a>
        </div>
        <nav
          className="header__nav"
        >
          <ul
            className="header__nav-list"
          >
            <li
              className="header__nav-item user"
            >
              <a
                className="header__nav-link header__nav-link--profile"
                href="/favorites"
                onClick={[Function]}
              >
                <div
                  className="header__avatar-wrapper user__avatar-wrapper"
                >
                  <img
                    src="https://4.react.pages.academy/six-cities/img/john_doe.jpg"
                    style={
                      Object {
                        "borderRadius": "50%",
                      }
                    }
                  />
                </div>
                <span
                  className="header__user-name user__name"
                >
                  mail@email.com
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
  <main
    className="page__main page__main--index null"
  >
    <h1
      className="visually-hidden"
    >
      Cities
    </h1>
    <div
      className="tabs"
    >
      <section
        className="locations container"
      >
        <ul
          className="locations__list tabs__list"
        >
          <li
            className="locations__item"
          >
            <a
              className="locations__item-link tabs__item tabs__item--active"
              href="#"
              onClick={[Function]}
            >
              <span>
                Amsterdam
              </span>
            </a>
          </li>
          <li
            className="locations__item"
          >
            <a
              className="locations__item-link tabs__item null"
              href="#"
              onClick={[Function]}
            >
              <span>
                Berlin
              </span>
            </a>
          </li>
        </ul>
      </section>
    </div>
    <div
      className="cities"
    >
      <div
        className="cities__places-container container"
      >
        <section
          className="cities__places places"
        >
          <h2
            className="visually-hidden"
          >
            Places
          </h2>
          <b
            className="places__found"
          >
            2
             places to stay in 
            Amsterdam
          </b>
          <form
            action="#"
            className="places__sorting"
            method="get"
          >
            <span
              className="places__sorting-caption"
            >
              Sort by 
            </span>
            <span
              className="places__sorting-type"
              onClick={[Function]}
              style={
                Object {
                  "outline": "none",
                  "userSelect": "none",
                }
              }
              tabIndex={0}
            >
              Popular
              <svg
                className="places__sorting-arrow"
                height={4}
                width={7}
              >
                <use
                  xlinkHref="#icon-arrow-select"
                />
              </svg>
            </span>
          </form>
          <div
            className="cities__places-list places__list tabs__content"
          >
            <article
              className="place-card cities__place-card"
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
            >
              <div
                className="place-card__image-wrapper cities__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height={200}
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg"
                    width={260}
                  />
                </a>
              </div>
              <div
                className="place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €248
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <a
                    className="place-card__bookmark-button button"
                    href="/"
                    onClick={[Function]}
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height={19}
                      width={18}
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </a>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "80%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating 
                      4.4
                    </span>
                  </div>
                </div>
                <a
                  href="/offer/1"
                  onClick={[Function]}
                >
                  <h2
                    className="place-card__name"
                  >
                    Penthouse, 4-5 rooms + 5 balconies
                  </h2>
                </a>
                <p
                  className="place-card__type"
                >
                  hotel
                </p>
              </div>
            </article>
            <article
              className="place-card cities__place-card"
              onMouseLeave={[Function]}
              onMouseOver={[Function]}
            >
              <div
                className="place-card__image-wrapper cities__image-wrapper"
              >
                <a
                  href="#"
                >
                  <img
                    alt="Place image"
                    className="place-card__image"
                    height={200}
                    src="https://htmlacademy-react-2.appspot.com/six-cities/static/hotel/14.jpg"
                    width={260}
                  />
                </a>
              </div>
              <div
                className="place-card__info"
              >
                <div
                  className="place-card__price-wrapper"
                >
                  <div
                    className="place-card__price"
                  >
                    <b
                      className="place-card__price-value"
                    >
                      €248
                    </b>
                    <span
                      className="place-card__price-text"
                    >
                      / night
                    </span>
                  </div>
                  <a
                    className="place-card__bookmark-button button"
                    href="/"
                    onClick={[Function]}
                  >
                    <svg
                      className="place-card__bookmark-icon"
                      height={19}
                      width={18}
                    >
                      <use
                        xlinkHref="#icon-bookmark"
                      />
                    </svg>
                    <span
                      className="visually-hidden"
                    >
                      To bookmarks
                    </span>
                  </a>
                </div>
                <div
                  className="place-card__rating rating"
                >
                  <div
                    className="place-card__stars rating__stars"
                  >
                    <span
                      style={
                        Object {
                          "width": "80%",
                        }
                      }
                    />
                    <span
                      className="visually-hidden"
                    >
                      Rating 
                      4.4
                    </span>
                  </div>
                </div>
                <a
                  href="/offer/2"
                  onClick={[Function]}
                >
                  <h2
                    className="place-card__name"
                  >
                    Penthouse, 4-5 rooms + 5 balconies
                  </h2>
                </a>
                <p
                  className="place-card__type"
                >
                  hotel
                </p>
              </div>
            </article>
          </div>
        </section>
        <div
          className="cities__right-section"
        >
          <section
            className="map property__map"
            id="map"
          />
        </div>
      </div>
    </div>
  </main>
</div>
`;
